ballerina/websocket_service.bal

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/io;
+import ballerina/websocket;
+
+import graphql.parser;
+
+isolated service class WsService {
+    *websocket:Service;
+
+    private final Engine engine;
+    private final readonly & __Schema schema;
+    private final Context context;
+    private map<[websocket:Caller, parser:OperationNode]> clientsMap;
+    
+    isolated function init(Engine engine, __Schema schema) {
+        self.engine = engine;
+        self.schema = schema.cloneReadOnly();
+        self.context = new();
+        self.clientsMap = {};
+    }
+                
+    isolated remote function onClose(websocket:Caller caller, int statusCode, string reason) {
+        lock {
+            _ = self.clientsMap.remove(caller.getConnectionId());
+        }
+    }
+
+    isolated remote function onTextMessage(websocket:Caller caller, string data) returns websocket:Error? {
+        lock {
+            parser:OperationNode|ErrorDetail node = validateSubscriptionPayload(data, self.engine);
+            if node is parser:OperationNode {
+                self.clientsMap[caller.getConnectionId()] = [caller, node];
+
+                stream<any, io:Error?> sourceStream = new(); // receiving stream from "getSubscriptionResponse(...)" -> Not implemented yet
+                any|io:Error? next = sourceStream.iterator().next();
+
+                while next !is io:Error? {
+                    check broadcast(self.engine, self.schema, self.context, next, self.clientsMap);
```

Comment:
What happens when this returns an error? I think we should handle it internally and send a proper GraphQL error response.

Reply:
`broadcast()` is returning a `websocket:Error` like in line 57. This error is the return type of `writeTextMessage()` function in websocket. Do we need to handle it internally?

Reply:
Let's keep this as it is for now then. But we should look at how other libraries handle these kinds of scenarios. I think we should send a GraphQL specific error here.